### ISSUE TITLE
`edd_is_checkout()` is slow

### ISSUE THREAD

**Issue by zackkatz** [awesomemotive/easy-digital-downloads#9741]
## Problem: `edd_is_checkout()` is slow and is called many times

`edd_is_checkout()` is called more than 20 times per page load. That wouldn't be a problem, except the function is slow: it parses the post content of the current page to check for embedded shortcodes and blocks, and it also has a filter. This becomes a problem for page speed.

### Function callers on a single page load

```
/wp-content/plugins/easy-digital-downloads-pro/includes/plugin-compatibility.php@124
/wp-content/plugins/easy-digital-downloads-pro/includes/checkout/functions.php@291
/wp-content/plugins/easy-digital-downloads-pro/includes/checkout/functions.php@312
/wp-content/plugins/easy-digital-downloads-pro/includes/payments/actions.php@348
/wp-content/plugins/easy-digital-downloads-pro/includes/plugin-compatibility.php@41
/wp-content/plugins/edd-software-licensing/includes/license-renewals.php@44
/wp-content/plugins/easy-digital-downloads-pro/includes/blocks/includes/checkout/forms.php@146
/wp-content/plugins/easy-digital-downloads-pro/includes/scripts.php@114
/wp-content/plugins/easy-digital-downloads-pro/includes/scripts.php@156
/wp-content/plugins/easy-digital-downloads-pro/includes/misc-functions.php@426
/wp-content/plugins/easy-digital-downloads-pro/includes/scripts.php@206
/wp-content/plugins/easy-digital-downloads-pro/includes/checkout/functions.php@74
/wp-content/plugins/easy-digital-downloads-pro/src/Pro/Checkout/GeoIP.php@153
/wp-content/plugins/easy-digital-downloads-pro/includes/template-functions.php@824
/wp-content/plugins/edd-software-licensing/src/Assets/Loader.php@50
/wp-content/plugins/edd-software-licensing/includes/scripts.php@89
/wp-content/plugins/edd-recurring/edd-recurring.php@1675
/wp-content/plugins/easy-digital-downloads-pro/includes/gateways/paypal/scripts.php@93
/wp-content/plugins/easy-digital-downloads-pro/includes/gateways/stripe/includes/scripts.php@36
/wp-content/plugins/easy-digital-downloads-pro/includes/gateways/stripe/includes/scripts.php@124
/wp-content/plugins/easy-digital-downloads-pro/includes/template-functions.php@782
/wp-content/plugins/easy-digital-downloads-pro/includes/checkout/functions.php@568
```

## A potential fix: Apply static caching

By applying a simple static cache, the function fully processes once per request.

Please see incoming pull request.

### Stats

These are numbers generated using the Code Profiler Pro plugin:

- Before: **0.234525 seconds**
- After: **0.031655**

### Potential downsides to this fix:

Since the filters are currently being run after each call, and the function is called in so many places, perhaps a plugin may want to modify a response from `is_checkout()` inside one of the callers and not another.

I don't think this is consideration holds up, though:

1. Either the page is checkout or it isn't. That shouldn't be modified mid-request.
2. If a filter wants to modify the output of `is_checkout()`, ideally that filter should be added before the first use of the function.

**Comment by robincornett** [comment]
Hey @zackkatz! Thank you for reporting on this and for submitting a PR. We are working to improve this check in our next EDD release, and I think we can combine the changes we are making with your suggestions and still be fully compatible with filter usage. We appreciate you!

**Comment by cklosowski** [comment]
@zackkatz This was resolved in EDD 3.1.1

**Comment by zackkatz** [comment]
Great! Thank you 🙏
